daemon/logs.go

```diff
@@ -14,13 +13,44 @@ import (
 	timetypes "github.com/docker/docker/api/types/time"
 	"github.com/docker/docker/container"
 	"github.com/docker/docker/daemon/logger"
-	"github.com/docker/docker/pkg/ioutils"
-	"github.com/docker/docker/pkg/stdcopy"
 )
 
-// ContainerLogs hooks up a container's stdout and stderr streams
-// configured with the given struct.
-func (daemon *Daemon) ContainerLogs(ctx context.Context, containerName string, config *backend.ContainerLogsConfig, started chan struct{}) error {
+// ContainerLogs copies the container's log channel to the channel provided in
+// the config. If ContainerLogs returns an error, no messages have been copied.
+// and the channel will be closed without data.
+//
+// if it returns nil, the config channel will be active and return log
+// messages until it runs out or the context is canceled.
+func (daemon *Daemon) ContainerLogs(ctx context.Context, containerName string, config *backend.ContainerLogsConfig) (rerr error) {
+
+	// O K so every path out of this function is an error before we start the
+	// streaming goroutine at the bottom, and nil after. if we return an error,
+	// that goroutine doesn't start (and doesn't close this channel, which, as
+	// the writer, it's our responsibility to close).
+	defer func() {
+		if rerr != nil {
+			close(config.Messages)
```

Comment:
I'd try to avoid closing a channel that's passed in. It means, for example, that you can't call the function twice on the same `ContainerLogsConfig`.

I think it's better to create a channel in this function and return it. I saw this was discussed in some of the earlier comments.